### PR TITLE
chore: Bump go-jose dependency and mod tidy

### LINKED ISSUE #417: Upgrade go-jose to fix CVE-2024-28180
Two weeks ago github.com/go-jose/go-jose released [v.4.0.1](https://github.com/go-jose/go-jose/releases/tag/v4.0.1) to fix a  published vulnerability ([CVE-2024-28180](https://www.cve.org/CVERecord?id=CVE-2024-28180).

This is issue is to bump the dependency from v3.0.1 to v4.0.1.

(I can do a pull request with the bump)

### PR DESCRIPTION
**Pull request by bcandeias**
This bumps the go-jose dependency to a version that addresses vulnerability CVE-2024-28180.

Tested with the below:

```
$ go fmt ./... ; go mod tidy ; go build ./... ; go test ./...
go: finding module for package github.com/go-jose/go-jose/v3
go: downloading github.com/go-jose/go-jose/v3 v3.0.3
go: downloading github.com/go-jose/go-jose v2.6.3+incompatible
go: found github.com/go-jose/go-jose/v3 in github.com/go-jose/go-jose/v3 v3.0.3
go: downloading github.com/stretchr/testify v1.8.2
go: downloading golang.org/x/crypto v0.19.0
Run 'go help' for usage.
?   	github.com/coreos/go-oidc/v3/example/idtoken	[no test files]
?   	github.com/coreos/go-oidc/v3/example/userinfo	[no test files]
ok  	github.com/coreos/go-oidc/v3/oidc	1.528s
```

Closes #417

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,17 +1,19 @@
 module github.com/coreos/go-oidc/v3
 
-go 1.19
+go 1.21
```
Comment (ericchiang) [inline go.mod:3]: To the best of my knowledge, this directive describes "this package can only be used with go 1.21 and up." But this package works fine with older versions

File: go.mod [hunk 2]
```diff
@@ -1,17 +1,19 @@
...
+
+toolchain go1.21.5
```
Comment (ericchiang) [inline go.mod:5]: Wouldn't expect us to pin a specific toolchain version like this

**Review by ericchiang — COMMENTED** [review]
Don't know why the toolchain directive was added, but I'm not seeing that locally

```
% go get github.com/go-jose/go-jose/v3@v3.0.3
go: downloading github.com/go-jose/go-jose/v3 v3.0.3
go: downloading golang.org/x/crypto v0.19.0
go: upgraded github.com/go-jose/go-jose/v3 v3.0.1 => v3.0.3
go: upgraded golang.org/x/crypto v0.14.0 => v0.19.0
% go mod tidy
go: downloading golang.org/x/net v0.17.0
go: downloading google.golang.org/appengine v1.6.8
% cat go.mod 
module github.com/coreos/go-oidc/v3

go 1.19

require (
	github.com/go-jose/go-jose/v3 v3.0.3
	golang.org/x/net v0.17.0
	golang.org/x/oauth2 v0.13.0
)

require (
	github.com/golang/protobuf v1.5.3 // indirect
	golang.org/x/crypto v0.19.0 // indirect
	google.golang.org/appengine v1.6.8 // indirect
	google.golang.org/protobuf v1.31.0 // indirect
)
```

**Comment by bcandeias** [comment]
@ericchiang sorry but I could only get to this now. You're right, I let these changes slide in with my local setup.
In the meantime I saw you did bump to 4.
Thanks, and apologies for not being more of a help 😅 
Closing this PR.

**Pull request closed without merging**